### ISSUE TITLE
❓ [Question] Not faster vs torch::jit 

### ISSUE THREAD

**Issue by JuncFang-git** [NVIDIA/Torch-TensorRT#586]
## ❓ Question
I run my model used TrTorch and torch::jit both on fp16 with C++ API,  but Trtorch is not faster than JIT.
What can I do to get the reason? 

Some information may be helpful. 
1. I used two plugins to just call the libtorch function (inverse and grid_smapler).
2. I fix some bugs by change the pytorch model code discript at #585 #584 
3. I compile trtorch model and run it all with C++ API.
4. the compile code is :
```
std::cout << "Load ts detector model ..." << std::endl;
    torch::jit::Module ts_detector_model;
    try{
        // Deserialize the ScriptModule from a file using torch::jit::load().
        ts_detector_model = torch::jit::load(TS_DETECTOR_PATH);
    }
    catch (const c10::Error& e){
        std::cerr << "Error loading the model \n";
        return -1;
    }

    // convert trt detector
    std::cout << "Convert trt detector model ..." << std::endl;
    ts_detector_model.to(at::kCUDA);
    ts_detector_model.eval();

    std::vector<trtorch::CompileSpec::Input> inputs_d = {
      trtorch::CompileSpec::Input(std::vector<int64_t>({1, 3, 256, 256}), trtorch::CompileSpec::DataType::kHalf)};
    auto info_d = trtorch::CompileSpec(inputs_d);
    info_d.enabled_precisions.insert(trtorch::CompileSpec::DataType::kHalf);

    auto trt_detector_model = trtorch::CompileGraph(ts_detector_model, info_d);
   
   // generator complied like above
   ......
```

the runtim code :
```
torch::jit::Module trt_detector_model;
    try{
        // Deserialize the ScriptModule from a file using torch::jit::load().
        trt_detector_model = torch::jit::load(TRT_DETECTOR_PATH);
    }
    catch (const c10::Error& e){
        std::cerr << "error loading the model \n";
        return -1;
    }
    trt_detector_model.to(at::kCUDA);
    trt_detector_model.eval();

    torch::jit::Module trt_generator_model;
    try{
        // Deserialize the ScriptModule from a file using torch::jit::load().
        trt_generator_model = torch::jit::load(TRT_GENERATOR_PATH);
    }
    catch (const c10::Error& e){
        std::cerr << "error loading the model \n";
        return -1;
    }
    trt_generator_model.to(at::kCUDA);
    trt_generator_model.eval();

    std::cout << "Run trt model ... " << std::endl;
    auto in0 = torch::ones({1, 3, 256, 256}, {torch::kCUDA}).to(torch::kFloat16);
    std::cout << "Run detector ... " << std::endl;
    auto out0_ = trt_detector_model.forward({in0});
    auto out0 = out0_.toTuple()->elements()[1].toTensor();
    std::cout << "====\tdetector out mean and std\t====" << std::endl;
    std::cout << at::mean(out0) << "\n" << at::std(out0) << std::endl;

    auto in1 = torch::ones({1, 3, 256, 256}, {torch::kCUDA}).to(torch::kFloat16);
    auto in2 = torch::ones({1, 10, 2}, {torch::kCUDA}).to(torch::kFloat16);
    auto in3 = torch::ones({1, 10, 2}, {torch::kCUDA}).to(torch::kFloat16);
    auto in4 = torch::ones({1, 10, 2, 2}, {torch::kCUDA}).to(torch::kFloat16);
    std::cout << "Run generator ... " << std::endl;
    auto out1 = trt_generator_model.forward({in1, in2, out0.to(torch::kFloat16), in3, in4}).toTensor();
    std::cout << "====\tgenerator out mean and std\t====" << std::endl;
```

## Environment

> Build information about the TRTorch compiler can be found by turning on debug messages

> I build the Trtorch recently(21.8.19) use the master branch.

 - TRTorch Version (8.0.1.6): 
 - PyTorch Version (libtorch 1.9.0):
 - OS (Ubuntu):
 - How you installed PyTorch (`libtorch):
 - Build command you used (bazel build //:libtrtorch --compilation_mode opt):
 - Are you using local sources or building from archives:
 - Python version: 3.8
 - CUDA version: 11.1
 - GPU models and configuration: GTX3070

**Comment by itsliupeng** [comment]
Your detection model may have many tensor indexes and slice operations, that can emit multiple Torchscript IR, and those IRs may not be merged and fused well in TRTorch converters and during TensorRT building.
I suggest that you can use `nsys profile` timeline to seek which PyTorch layer generates multiple TensorRT layers,  such as tensor index `a[:,:5,:,:]`.

**Comment by JuncFang-git** [comment]
> Your detection model may have many tensor indexes and slice operations, that can emit multiple Torchscript IR, and those IRs may not be merged and fused well in TRTorch converters and during TensorRT building.
> I suggest that you can use `nsys profile` timeline to seek which PyTorch layer generates multiple TensorRT layers, such as tensor index `a[:,:5,:,:]`.

Hi @itsliupeng ,
Thanks for your advice！nsys is a nice tool. 👍 
I find the bug may be caused by my plugins(inverse and grid_sampler ), I write these two plugins to just call a libtorch function like  the official example [https://github.com/NVIDIA/TRTorch/blob/master/core/plugins/impl/interpolate_plugin.cpp](url)

But there are some bugs:
1. plugins are slower than the pytorch function.
pytorch inverse :   0.427 ms
trtorch inverse plugin: 3.081ms
pytorch grid_sampler: 0.034ms
trtorch grid_smapler plugin: 0.347 ms
Analyzed by "torch profiler":
![image](https://user-images.githubusercontent.com/76929740/130341310-adace78d-cc88-4b63-80e7-6f5b3a442ca1.png)
![image](https://user-images.githubusercontent.com/76929740/130341328-aa2a2501-41e8-47b1-99c8-3b9a45078b39.png)
Analyzed by "nsys"":
![image](https://user-images.githubusercontent.com/76929740/130341386-c618faa8-a0a6-49ab-a192-b7d4bb77f0d1.png)
![image](https://user-images.githubusercontent.com/76929740/130341405-ee80eacc-1e47-4dfe-b2e1-4d3194b533c3.png)

2. inverse plugin take a huge time at some loop step:
![image](https://user-images.githubusercontent.com/76929740/130341489-8bfad487-3c86-404e-973c-0eaf236dcc3f.png)

So,
1. Is the plugin just call libtorch function will take more time as above? 
2. How can I fix the second bug?

**Comment by JuncFang-git** [comment]
Hi,
I rewrite a 2dims inverse converter using  the tensorrt operation to replace the inverse plugin, but it still costs lots of time caused by the "output copy".
![image](https://user-images.githubusercontent.com/76929740/130450855-da0c435d-fd36-433d-b928-2a015f15b7ce.png)
this is my code:
`auto torchinverse TRTORCH_UNUSED = RegisterNodeConversionPatterns().pattern(
    {
        "aten::inverse(Tensor self) -> (Tensor)",
        [](ConversionCtx* ctx, const torch::jit::Node* n, args& args) -> bool {
            /*
             * This inverse converter just for the matrix with the shape like (B, C, 2, 2)
             * FUCTION:
             * M = [[a, b],
             *      [c, d]], 
             * M_adjoint = [[d, -b],
             *              [-c, a]]
             * M_inverse = M_adjoint / M_det = M_adjoint / (ad - bc)
             */

            auto in = args[0].ITensorOrFreeze(ctx);
            

            // flatten
            auto start_dim = 2;
            auto end_dim = 3;
            auto in_shape = util::toVec(in->getDimensions());
            std::vector<int64_t> flatten_out_shape;
            flatten_out_shape = torch::flatten(torch::rand(in_shape), start_dim, end_dim).sizes().vec();

            auto shuffle = ctx->net->addShuffle(*in);
            TRTORCH_CHECK(shuffle, "Unable to create shuffle layer from node: " << *n);
            shuffle->setReshapeDimensions(util::toDims(flatten_out_shape));
            auto flattened = shuffle->getOutput(0);
            LOG_DEBUG("Inverse(Juncfang) flattened tensor shape: " << flattened->getDimensions());

            //// Get M_adjoint
            // chagne element position
            at::Tensor indices = torch::tensor({3, 1, 2, 0}).to(torch::kI32);
            auto indicesTensor = tensor_to_const(ctx, indices);

            auto gather_layer = ctx->net->addGather(*flattened, *indicesTensor, 2);
            TRTORCH_CHECK(gather_layer, "Unable to create gather layer from node: " << *n);
            auto gather_out = gather_layer->getOutput(0);
            LOG_DEBUG("Inverse(Juncfang) gather_out tensor shape: " << gather_out->getDimensions());

            // multiply weight
            at::Tensor w = torch::tensor({1, -1, -1, 1}).to(torch::kF16); //TODO to input type
            auto wTensor = tensor_to_const(ctx, w);

            auto multiply_layer = add_elementwise(ctx, nvinfer1::ElementWiseOperation::kPROD, gather_out, wTensor, util::node_info(n)+"_prod");
            auto multiply_out = multiply_layer->getOutput(0);
            LOG_DEBUG("Inverse(Juncfang) multiply_out tensor shape: " << multiply_out->getDimensions());

            // reshape to get M_adjoint
            auto reshape_layer = ctx->net->addShuffle(*multiply_out);
            reshape_layer->setReshapeDimensions(in->getDimensions());
            auto M_adjoint = reshape_layer->getOutput(0);
            LOG_DEBUG("Inverse(Juncfang) M_adjoint tensor shape: " << M_adjoint->getDimensions());


            //// Get M_det
            // get a*d
            indices = torch::tensor({0, 3}).to(torch::kI32);
            indicesTensor = tensor_to_const(ctx, indices);
            auto gather_layer1 = ctx->net->addGather(*flattened, *indicesTensor, 2);
            auto gather_layer_out1 = gather_layer1->getOutput(0);
            auto axis_mask1 = 1 << 2;
            auto prod_layer1 = ctx->net->addReduce(*gather_layer_out1, nvinfer1::ReduceOperation::kPROD, axis_mask1, false);
            auto pord1 = prod_layer1->getOutput(0);
            LOG_DEBUG("Inverse(Juncfang) pord1 tensor shape: " << pord1->getDimensions());

            // get b*c
            indices = torch::tensor({1, 2}).to(torch::kI32);
            indicesTensor = tensor_to_const(ctx, indices);
            auto gather_layer2 = ctx->net->addGather(*flattened, *indicesTensor, 2);
            auto gather_layer_out2 = gather_layer2->getOutput(0);
            auto axis_mask2 = 1 << 2;
            auto prod_layer2 = ctx->net->addReduce(*gather_layer_out2, nvinfer1::ReduceOperation::kPROD, axis_mask2, false);
            auto pord2 = prod_layer2->getOutput(0);
            LOG_DEBUG("Inverse(Juncfang) pord2 tensor shape: " << pord2->getDimensions());

            // a*d - b*c
            auto sub_layer = ctx->net->addElementWise(*pord1, *pord2, nvinfer1::ElementWiseOperation::kSUB);
            auto M_det = sub_layer->getOutput(0);
            LOG_DEBUG("Inverse(Juncfang) M_det tensor shape: " << M_det->getDimensions());

            //// Get M_inverse
            // Expand dims
            auto detDim = util::toVec(M_det->getDimensions());
            auto adjointDim = util::toVec(M_adjoint->getDimensions());
            nvinfer1::Dims dims;
            dims.nbDims = adjointDim.size();
            for (size_t i = 0; i < adjointDim.size(); i++){
                dims.d[i] = i < detDim.size() ? detDim[i] : 1 ;
            }
            auto detShuffle = ctx->net->addShuffle(*M_det);
            detShuffle->setReshapeDimensions(dims);
            M_det = detShuffle->getOutput(0);
            LOG_DEBUG("Inverse(Juncfang) M_det tensor shape: " << M_det->getDimensions());
            auto out_layer = add_elementwise(ctx, nvinfer1::ElementWiseOperation::kDIV, M_adjoint, M_det, util::node_info(n)+"_div");
            out_layer->setName(util::node_info(n).c_str());

            auto out_tensor = ctx->AssociateValueAndTensor(n->outputs()[0], out_layer->getOutput(0));
            LOG_DEBUG("Inverse(Juncfang) Output tensor shape: " << out_tensor->getDimensions());
            // return false; // stop point for debug 
        }
    }
);`
Can anyone give me some advice? @narendasan @itsliupeng @chengyute

**Comment by narendasan** [comment]
Haven't looked too deeply at this but, to answer an earlier question, yes you can expect overhead calling libtorch from a TensorRT plugin.
